### PR TITLE
Add experimental SO2 equivariant convolution and activation

### PR DESCRIPTION
**Pull request by laserkelvin**
# PhysicsNeMo Pull Request

## Description

This PR adds a new `symmetry` namespace within `experimental`, which is intended to house some performant implementations of operations that preserve physical symmetries. What's included in this current PR is support for SO(2) transformations (`SO2Convolution`), and an accompanying rotational equivariance preserving nonlinearity, `GateActivation`. The former is written to take advantage of `einsum`s to efficiently (and correctly) compute linear transforms in a spherical harmonic basis. The new unit tests added include both conventional tests (e.g. expected output shapes), as well as some to check for 2D equivariance.

These layers are nominally general purpose (i.e. any point cloud), but are derived (with heavy modifications to data layout and subsequently performance) from the [`fairchem`](https://github.com/facebookresearch/fairchem/tree/main) repository.

@peterdsharpe this is more or less the same as what you've reviewed already, and @CharlelieLrt please review as much as you want, but including you because this adds new stuff into `physicsnemo.experimental.nn` and figured you might want to monitor for model coding standards.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [x] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Dependencies

None

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/experimental/nn/symmetry/activation.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+        x: Float[Tensor, "batch lmax_plus_1 mmax_plus_1 2 channels_with_gates"],
+    ) -> Float[Tensor, "batch lmax_plus_1 mmax_plus_1 2 channels"]:
+        r"""Apply gated activation to equivariant features.
+
+        Gates are extracted from the input tensor itself, embedded at the
+        (l=0, m=0, real=0) position in the channel dimension.
+
+        Parameters
+        ----------
+        x : Float[Tensor, "batch lmax_plus_1 mmax_plus_1 2 channels_with_gates"]
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply physicsnemo/experimental/nn/symmetry/activation.py]: This resolves correctly in unit tests, as `physicsnemo.nn.module.activations.get_activations` looks into an `ACT2FN` mapping

File: physicsnemo/experimental/nn/symmetry/activation.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+from physicsnemo.experimental.nn.symmetry.grid import make_grid_mask
+from physicsnemo.nn import get_activation
+
+__all__ = [
+    "GateActivation",
+]
+
+
+class GateActivation(nn.Module):
```
Comment (CharlelieLrt) [inline physicsnemo/experimental/nn/symmetry/activation.py]: If I recall we agreed that everything should be direct subclasses of `physicsnemo.Module`, even layers. So, no subclass of `torch.nn.Module` allowed.

File: physicsnemo/experimental/nn/symmetry/activation.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+    ----------
+    lmax : int
+        Maximum spherical harmonic degree. Must be >= 1 (need at least one
+        gated degree for this activation to be meaningful).
+    mmax : int
+        Maximum spherical harmonic order. Must satisfy 0 <= mmax <= lmax.
+    channels : int
+        Number of output feature channels (not including gates).
+
+    Attributes
```
Comment (CharlelieLrt) [inline physicsnemo/experimental/nn/symmetry/activation.py:80]: I think none of our models have an `Attributes` section in their docstring, but I have nothing against it. When moving this out of experimental into the core library, we should make sure that it renders correctly in the online docs

File: physicsnemo/experimental/nn/symmetry/activation.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+    tensor. The first ``channels`` values at this position are the l=0 features,
+    and the remaining ``lmax * channels`` values are the gates for l>0 degrees.
+
+    Output tensor layout: [batch, lmax+1, mmax+1, 2, channels]
+        - Same structure as input but without the gate channels
+        - Gate channels are consumed during the activation, not passed through
+
+    See Also
+    --------
+    SO2Convolution - SO2-preserving linear transformation layer.
```
Comment (CharlelieLrt) [inline physicsnemo/experimental/nn/symmetry/activation.py:124]: nit: I think for links within the API docs, the right syntax is, for example, :class:`physicsnemo.experimental.[...].SO2Convolution`. Nothing critical, since this model is in experimental

File: physicsnemo/experimental/nn/symmetry/activation.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        return self.lmax * self.channels
+
+    @property
+    def total_in_channels(self) -> int:
+        """Total input channels including embedded gates."""
+        return self.channels + self.gate_channels
+
+    def forward(
+        self,
+        x: Float[Tensor, "batch lmax_plus_1 mmax_plus_1 2 channels_with_gates"],
```
Comment (CharlelieLrt) [inline physicsnemo/experimental/nn/symmetry/activation.py:211]: Most other models I have seen use some shorter names for batches ("B") and for channels ("C")

File: physicsnemo/experimental/nn/symmetry/activation.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+    ) -> Float[Tensor, "batch lmax_plus_1 mmax_plus_1 2 channels"]:
+        r"""Apply gated activation to equivariant features.
```
Comment (CharlelieLrt) [inline physicsnemo/experimental/nn/symmetry/activation.py]: In our docs we put the docstring of the forward in the class docstring, under a `Forward` section. The output of the forward method then goes also in the class docstring in a section called `Outputs`. There is no need to repeat the docstring in the method itself.

File: physicsnemo/experimental/nn/symmetry/so2_conv.py [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+
+        # Add LayerNorm + SiLU for all but the last layer
+        if i < len(channels_list) - 2:
+            modules.append(nn.LayerNorm(out_ch))
+            modules.append(nn.SiLU())
+
+    return nn.Sequential(*modules)
+
+
+class SO2Convolution(nn.Module):
```
Comment (CharlelieLrt) [inline physicsnemo/experimental/nn/symmetry/so2_conv.py]: Mostly the same comments as for the class `GateActivation` (direct subclass of `physicsnemo.Module`, docstring formatting ,etc.)

File: physicsnemo/experimental/nn/symmetry/activation.py (hunk 2 above)
Comment (laserkelvin) [reply physicsnemo/experimental/nn/symmetry/activation.py]: Addressed in 828e956

File: physicsnemo/experimental/nn/symmetry/activation.py (hunk 3 above)
Comment (laserkelvin) [reply physicsnemo/experimental/nn/symmetry/activation.py:80]: Sounds good - since it's experimental for now and we don't have proper docs, I'm inclined to leave it in for now

File: physicsnemo/experimental/nn/symmetry/activation.py (hunk 4 above)
Comment (laserkelvin) [reply physicsnemo/experimental/nn/symmetry/activation.py:124]: Actually I think it works either way, but it might depend on how `sphinx` is set up. IMO we come back to this when we move out of experimental, when the API is finalized

File: physicsnemo/experimental/nn/symmetry/activation.py (hunk 5 above)
Comment (laserkelvin) [reply physicsnemo/experimental/nn/symmetry/activation.py:211]: I can change it, but do you think it'll look weird? `B lmax_plus_1 mmax_plus_1 2 C`

File: physicsnemo/experimental/nn/symmetry/activation.py (hunk 6 above)
Comment (laserkelvin) [reply physicsnemo/experimental/nn/symmetry/activation.py]: Addressed in 2dde1bd

**Comment by laserkelvin** [comment]
@CharlelieLrt I think I've addressed your comments, except the `jaxtyping` one. I'm happy to go either way, just let me know

File: physicsnemo/experimental/nn/symmetry/activation.py (hunk 5 above)
Comment (CharlelieLrt) [reply physicsnemo/experimental/nn/symmetry/activation.py:211]: Yes, it does look weird 😆 Let's keep it as `batch_size`
